File: okhttp-tests/src/test/java/okhttp3/internal/http/RouteSelectorTest.java
/*
 * Copyright (C) 2012 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package okhttp3.internal.http;

import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Proxy;
import java.net.ProxySelector;
import java.net.SocketAddress;
import java.net.URI;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.NoSuchElementException;
import javax.net.SocketFactory;
import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLSocketFactory;
import okhttp3.Address;
import okhttp3.Authenticator;
import okhttp3.ConnectionSpec;
import okhttp3.FakeDns;
import okhttp3.Protocol;
import okhttp3.Route;
import okhttp3.internal.RouteDatabase;
import okhttp3.internal.Util;
import okhttp3.internal.tls.SslClient;
import org.junit.Before;
import org.junit.Test;

import static java.net.Proxy.NO_PROXY;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public final class RouteSelectorTest {
  public final List<ConnectionSpec> connectionSpecs = Util.immutableList(
      ConnectionSpec.MODERN_TLS,
      ConnectionSpec.COMPATIBLE_TLS,
      ConnectionSpec.CLEARTEXT);

  private static final int proxyAPort = 1001;
  private static final String proxyAHost = "proxya";
  private static final Proxy proxyA =
      new Proxy(Proxy.Type.HTTP, InetSocketAddress.createUnresolved(proxyAHost, proxyAPort));
  private static final int proxyBPort = 1002;
  private static final String proxyBHost = "proxyb";
  private static final Proxy proxyB =
      new Proxy(Proxy.Type.HTTP, InetSocketAddress.createUnresolved(proxyBHost, proxyBPort));
  private String uriHost = "hosta";
  private int uriPort = 1003;

  private SocketFactory socketFactory;
  private final SslClient sslClient = SslClient.localhost();
  private final SSLSocketFactory sslSocketFactory = sslClient.socketFactory;
  private HostnameVerifier hostnameVerifier;

  private final Authenticator authenticator = Authenticator.NONE;
  private final List<Protocol> protocols = Arrays.asList(Protocol.HTTP_1_1);
  private final FakeDns dns = new FakeDns();
  private final RecordingProxySelector proxySelector = new RecordingProxySelector();
  private RouteDatabase routeDatabase = new RouteDatabase();

  @Before public void setUp() throws Exception {
    socketFactory = SocketFactory.getDefault();
    hostnameVerifier = HttpsURLConnection.getDefaultHostnameVerifier();
  }

  @Test public void singleRoute() throws Exception {
    Address address = httpAddress();
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);

    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(255, 1));
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(0), uriPort);
    dns.assertRequests(uriHost);

    assertFalse(routeSelector.hasNext());
    try {
      routeSelector.next();
      fail();
    } catch (NoSuchElementException expected) {
    }
  }

  @Test public void singleRouteReturnsFailedRoute() throws Exception {
    Address address = httpAddress();
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);

    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(255, 1));
    Route route = routeSelector.next();
    routeDatabase.failed(route);
    routeSelector = new RouteSelector(address, routeDatabase);
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(0), uriPort);
    assertFalse(routeSelector.hasNext());
    try {
      routeSelector.next();
      fail();
    } catch (NoSuchElementException expected) {
    }
  }

  @Test public void explicitProxyTriesThatProxysAddressesOnly() throws Exception {
    Address address = new Address(uriHost, uriPort, dns, socketFactory, null, null, null,
        authenticator, proxyA, protocols, connectionSpecs, proxySelector);
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);

    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(255, 2));
    assertRoute(routeSelector.next(), address, proxyA, dns.address(0), proxyAPort);
    assertRoute(routeSelector.next(), address, proxyA, dns.address(1), proxyAPort);

    assertFalse(routeSelector.hasNext());
    dns.assertRequests(proxyAHost);
    proxySelector.assertRequests(); // No proxy selector requests!
  }

  @Test public void explicitDirectProxy() throws Exception {
    Address address = new Address(uriHost, uriPort, dns, socketFactory, null, null, null,
        authenticator, NO_PROXY, protocols, connectionSpecs, proxySelector);
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);

    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(255, 2));
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(0), uriPort);
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(1), uriPort);

    assertFalse(routeSelector.hasNext());
    dns.assertRequests(uriHost);
    proxySelector.assertRequests(); // No proxy selector requests!
  }

  @Test public void proxySelectorReturnsNull() throws Exception {
    ProxySelector nullProxySelector = new ProxySelector() {
      @Override public List<Proxy> select(URI uri) {
        assertEquals(uriHost, uri.getHost());
        return null;
      }

      @Override public void connectFailed(
          URI uri, SocketAddress socketAddress, IOException e) {
        throw new AssertionError();
      }
    };

    Address address = new Address(uriHost, uriPort, dns, socketFactory, null, null, null,
        authenticator, null, protocols, connectionSpecs, nullProxySelector);
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);
    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(255, 1));
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(0), uriPort);
    dns.assertRequests(uriHost);

    assertFalse(routeSelector.hasNext());
  }

  @Test public void proxySelectorReturnsNoProxies() throws Exception {
    Address address = httpAddress();
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);

    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(255, 2));
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(0), uriPort);
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(1), uriPort);

    assertFalse(routeSelector.hasNext());
    dns.assertRequests(uriHost);
    proxySelector.assertRequests(address.url().uri());
  }

  @Test public void proxySelectorReturnsMultipleProxies() throws Exception {
    Address address = httpAddress();

    proxySelector.proxies.add(proxyA);
    proxySelector.proxies.add(proxyB);
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);
    proxySelector.assertRequests(address.url().uri());

    // First try the IP addresses of the first proxy, in sequence.
    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(255, 2));
    assertRoute(routeSelector.next(), address, proxyA, dns.address(0), proxyAPort);
    assertRoute(routeSelector.next(), address, proxyA, dns.address(1), proxyAPort);
    dns.assertRequests(proxyAHost);

    // Next try the IP address of the second proxy.
    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(254, 1));
    assertRoute(routeSelector.next(), address, proxyB, dns.address(0), proxyBPort);
    dns.assertRequests(proxyBHost);

    // Finally try the only IP address of the origin server.
    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(253, 1));
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(0), uriPort);
    dns.assertRequests(uriHost);

    assertFalse(routeSelector.hasNext());
  }

  @Test public void proxySelectorDirectConnectionsAreSkipped() throws Exception {
    Address address = httpAddress();

    proxySelector.proxies.add(NO_PROXY);
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);
    proxySelector.assertRequests(address.url().uri());

    // Only the origin server will be attempted.
    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(255, 1));
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(0), uriPort);
    dns.assertRequests(uriHost);

    assertFalse(routeSelector.hasNext());
  }

  @Test public void proxyDnsFailureContinuesToNextProxy() throws Exception {
    Address address = httpAddress();

    proxySelector.proxies.add(proxyA);
    proxySelector.proxies.add(proxyB);
    proxySelector.proxies.add(proxyA);
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);
    proxySelector.assertRequests(address.url().uri());

    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(255, 1));
    assertRoute(routeSelector.next(), address, proxyA, dns.address(0), proxyAPort);
    dns.assertRequests(proxyAHost);

    assertTrue(routeSelector.hasNext());
    dns.unknownHost();
    try {
      routeSelector.next();
      fail();
    } catch (UnknownHostException expected) {
    }
    dns.assertRequests(proxyBHost);

    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(255, 1));
    assertRoute(routeSelector.next(), address, proxyA, dns.address(0), proxyAPort);
    dns.assertRequests(proxyAHost);

    assertTrue(routeSelector.hasNext());
    dns.addresses(makeFakeAddresses(254, 1));
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(0), uriPort);
    dns.assertRequests(uriHost);

    assertFalse(routeSelector.hasNext());
  }

  @Test public void multipleProxiesMultipleInetAddressesMultipleConfigurations() throws Exception {
    Address address = httpsAddress();
    proxySelector.proxies.add(proxyA);
    proxySelector.proxies.add(proxyB);
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);

    // Proxy A
    dns.addresses(makeFakeAddresses(255, 2));
    assertRoute(routeSelector.next(), address, proxyA, dns.address(0), proxyAPort);
    dns.assertRequests(proxyAHost);
    assertRoute(routeSelector.next(), address, proxyA, dns.address(1), proxyAPort);

    // Proxy B
    dns.addresses(makeFakeAddresses(254, 2));
    assertRoute(routeSelector.next(), address, proxyB, dns.address(0), proxyBPort);
    dns.assertRequests(proxyBHost);
    assertRoute(routeSelector.next(), address, proxyB, dns.address(1), proxyBPort);

    // Origin
    dns.addresses(makeFakeAddresses(253, 2));
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(0), uriPort);
    dns.assertRequests(uriHost);
    assertRoute(routeSelector.next(), address, NO_PROXY, dns.address(1), uriPort);

    assertFalse(routeSelector.hasNext());
  }

  @Test public void failedRoutesAreLast() throws Exception {
    Address address = httpsAddress();
    RouteSelector routeSelector = new RouteSelector(address, routeDatabase);

    final int numberOfAddresses = 2;
    dns.addresses(makeFakeAddresses(255, numberOfAddresses));

    // Extract the regular sequence of routes from selector.
    List<Route> regularRoutes = new ArrayList<>();
    while (routeSelector.hasNext()) {
      regularRoutes.add(routeSelector.next());
    }

    // Check that we do indeed have more than one route.
    assertEquals(numberOfAddresses, regularRoutes.size());
    // Add first regular route as failed.
    routeDatabase.failed(regularRoutes.get(0));
    // Reset selector
    routeSelector = new RouteSelector(address, routeDatabase);

    List<Route> routesWithFailedRoute = new ArrayList<>();
    while (routeSelector.hasNext()) {
      routesWithFailedRoute.add(routeSelector.next());
    }

    assertEquals(regularRoutes.get(0),
        routesWithFailedRoute.get(routesWithFailedRoute.size() - 1));
    assertEquals(regularRoutes.size(), routesWithFailedRoute.size());
  }

  @Test public void getHostString() throws Exception {
    // Name proxy specification.
    InetSocketAddress socketAddress = InetSocketAddress.createUnresolved("host", 1234);
    assertEquals("host", RouteSelector.getHostString(socketAddress));
    socketAddress = InetSocketAddress.createUnresolved("127.0.0.1", 1234);
    assertEquals("127.0.0.1", RouteSelector.getHostString(socketAddress));

    // InetAddress proxy specification.
    socketAddress = new InetSocketAddress(InetAddress.getByName("localhost"), 1234);
    assertEquals("127.0.0.1", RouteSelector.getHostString(socketAddress));
    socketAddress = new InetSocketAddress(
        InetAddress.getByAddress(new byte[] {127, 0, 0, 1}), 1234);
    assertEquals("127.0.0.1", RouteSelector.getHostString(socketAddress));
    socketAddress = new InetSocketAddress(
        InetAddress.getByAddress("foobar", new byte[] {127, 0, 0, 1}), 1234);
    assertEquals("127.0.0.1", RouteSelector.getHostString(socketAddress));
  }

  private void assertRoute(Route route, Address address, Proxy proxy, InetAddress socketAddress,
      int socketPort) {
    assertEquals(address, route.address());
    assertEquals(proxy, route.proxy());
    assertEquals(socketAddress, route.socketAddress().getAddress());
    assertEquals(socketPort, route.socketAddress().getPort());
  }

  /** Returns an address that's without an SSL socket factory or hostname verifier. */
  private Address httpAddress() {
    return new Address(uriHost, uriPort, dns, socketFactory, null, null, null, authenticator, null,
        protocols, connectionSpecs, proxySelector);
  }

  private Address httpsAddress() {
    return new Address(uriHost, uriPort, dns, socketFactory, sslSocketFactory,
        hostnameVerifier, null, authenticator, null, protocols, connectionSpecs, proxySelector);
  }

  private static List<InetAddress> makeFakeAddresses(int prefix, int count) {
    try {
      List<InetAddress> result = new ArrayList<>();
      for (int i = 0; i < count; i++) {
        result.add(InetAddress.getByAddress(
            new byte[] {(byte) prefix, (byte) 0, (byte) 0, (byte) i}));
      }
      return result;
    } catch (UnknownHostException e) {
      throw new AssertionError();
    }
  }
}

File: okhttp/src/jvmTest/java/okhttp3/internal/connection/RouteSelectorTest.kt
/*
 * Copyright (C) 2012 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package okhttp3.internal.connection

import java.io.IOException
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Proxy
import java.net.ProxySelector
import java.net.SocketAddress
import java.net.URI
import java.net.UnknownHostException
import javax.net.SocketFactory
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.HttpsURLConnection
import okhttp3.Address
import okhttp3.Authenticator
import okhttp3.Call
import okhttp3.ConnectionSpec
import okhttp3.EventListener
import okhttp3.FakeDns
import okhttp3.OkHttpClientTestRule
import okhttp3.Protocol
import okhttp3.Request
import okhttp3.Route
import okhttp3.internal.connection.RouteSelector.Companion.socketHost
import okhttp3.internal.http.RecordingProxySelector
import okhttp3.internal.platform.Platform.Companion.isAndroid
import okhttp3.testing.PlatformRule
import okhttp3.testing.PlatformVersion.majorVersion
import okhttp3.tls.internal.TlsUtil.localhost
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.fail
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.RegisterExtension

class RouteSelectorTest {
  @RegisterExtension
  val platform = PlatformRule()

  @RegisterExtension
  val clientTestRule = OkHttpClientTestRule()

  val connectionSpecs = listOf(
    ConnectionSpec.MODERN_TLS,
    ConnectionSpec.COMPATIBLE_TLS,
    ConnectionSpec.CLEARTEXT
  )

  private val uriHost = "hosta"
  private val uriPort = 1003
  private lateinit var call: Call
  private lateinit var socketFactory: SocketFactory
  private val handshakeCertificates = localhost()
  private val sslSocketFactory = handshakeCertificates.sslSocketFactory()
  private lateinit var hostnameVerifier: HostnameVerifier
  private val authenticator = Authenticator.NONE
  private val protocols = listOf(Protocol.HTTP_1_1)
  private val dns = FakeDns()
  private val proxySelector = RecordingProxySelector()
  private val routeDatabase = RouteDatabase()

  @BeforeEach fun setUp() {
    call = clientTestRule.newClient().newCall(
      Request.Builder()
        .url("https://$uriHost:$uriPort/")
        .build()
    )
    socketFactory = SocketFactory.getDefault()
    hostnameVerifier = HttpsURLConnection.getDefaultHostnameVerifier()
  }

  @Test fun singleRoute() {
    val address = httpAddress()
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    assertThat(routeSelector.hasNext()).isTrue
    dns[uriHost] = dns.allocate(1)
    val selection = routeSelector.next()
    assertRoute(selection.next(), address, Proxy.NO_PROXY, dns.lookup(uriHost, 0), uriPort)
    dns.assertRequests(uriHost)
    assertThat(selection.hasNext()).isFalse
    try {
      selection.next()
      fail<Any>()
    } catch (expected: NoSuchElementException) {
    }
    assertThat(routeSelector.hasNext()).isFalse
    try {
      routeSelector.next()
      fail<Any>()
    } catch (expected: NoSuchElementException) {
    }
  }

  @Test fun singleRouteReturnsFailedRoute() {
    val address = httpAddress()
    var routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    assertThat(routeSelector.hasNext()).isTrue
    dns[uriHost] = dns.allocate(1)
    var selection = routeSelector.next()
    val route = selection.next()
    routeDatabase.failed(route)
    routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    selection = routeSelector.next()
    assertRoute(selection.next(), address, Proxy.NO_PROXY, dns.lookup(uriHost, 0), uriPort)
    assertThat(selection.hasNext()).isFalse
    try {
      selection.next()
      fail<Any>()
    } catch (expected: NoSuchElementException) {
    }
    assertThat(routeSelector.hasNext()).isFalse
    try {
      routeSelector.next()
      fail<Any>()
    } catch (expected: NoSuchElementException) {
    }
  }

  @Test fun explicitProxyTriesThatProxysAddressesOnly() {
    val address = Address(
      uriHost = uriHost,
      uriPort = uriPort,
      dns = dns,
      socketFactory = socketFactory,
      sslSocketFactory = null,
      hostnameVerifier = null,
      certificatePinner = null,
      proxyAuthenticator = authenticator,
      proxy = proxyA,
      protocols = protocols,
      connectionSpecs = connectionSpecs,
      proxySelector = proxySelector
    )
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    assertThat(routeSelector.hasNext()).isTrue
    dns[proxyAHost] = dns.allocate(2)
    val selection = routeSelector.next()
    assertRoute(selection.next(), address, proxyA, dns.lookup(proxyAHost, 0), proxyAPort)
    assertRoute(selection.next(), address, proxyA, dns.lookup(proxyAHost, 1), proxyAPort)
    assertThat(selection.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isFalse
    dns.assertRequests(proxyAHost)
    proxySelector.assertRequests() // No proxy selector requests!
  }

  @Test fun explicitDirectProxy() {
    val address = Address(
      uriHost = uriHost,
      uriPort = uriPort,
      dns = dns,
      socketFactory = socketFactory,
      sslSocketFactory = null,
      hostnameVerifier = null,
      certificatePinner = null,
      proxyAuthenticator = authenticator,
      proxy = Proxy.NO_PROXY,
      protocols = protocols,
      connectionSpecs = connectionSpecs,
      proxySelector = proxySelector
    )
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    assertThat(routeSelector.hasNext()).isTrue
    dns[uriHost] = dns.allocate(2)
    val selection = routeSelector.next()
    assertRoute(selection.next(), address, Proxy.NO_PROXY, dns.lookup(uriHost, 0), uriPort)
    assertRoute(selection.next(), address, Proxy.NO_PROXY, dns.lookup(uriHost, 1), uriPort)
    assertThat(selection.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isFalse
    dns.assertRequests(uriHost)
    proxySelector.assertRequests() // No proxy selector requests!
  }

  /**
   * Don't call through to the proxy selector if we don't have a host name.
   * https://github.com/square/okhttp/issues/5770
   */
  @Test fun proxySelectorNotCalledForNullHost() {
    // The string '>' is okay in a hostname in HttpUrl, which does very light hostname validation.
    // It is not okay in URI, and so it's stripped and we get a URI with a null host.
    val bogusHostname = ">"
    val address = Address(
      uriHost = bogusHostname,
      uriPort = uriPort,
      dns = dns,
      socketFactory = socketFactory,
      sslSocketFactory = null,
      hostnameVerifier = null,
      certificatePinner = null,
      proxyAuthenticator = authenticator,
      proxy = null,
      protocols = protocols,
      connectionSpecs = connectionSpecs,
      proxySelector = proxySelector
    )
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    assertThat(routeSelector.hasNext()).isTrue
    dns[bogusHostname] = dns.allocate(1)
    val selection = routeSelector.next()
    assertRoute(selection.next(), address, Proxy.NO_PROXY, dns.lookup(bogusHostname, 0), uriPort)
    assertThat(selection.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isFalse
    dns.assertRequests(bogusHostname)
    proxySelector.assertRequests() // No proxy selector requests!
  }

  @Test fun proxySelectorReturnsNull() {
    val nullProxySelector: ProxySelector = object : ProxySelector() {
      override fun select(uri: URI): List<Proxy>? {
        assertThat(uri.host).isEqualTo(uriHost)
        return null
      }

      override fun connectFailed(uri: URI, socketAddress: SocketAddress, e: IOException) {
        throw AssertionError()
      }
    }

    val address = Address(
      uriHost = uriHost,
      uriPort = uriPort,
      dns = dns,
      socketFactory = socketFactory,
      sslSocketFactory = null,
      hostnameVerifier = null,
      certificatePinner = null,
      proxyAuthenticator = authenticator,
      proxy = null,
      protocols = protocols,
      connectionSpecs = connectionSpecs,
      proxySelector = nullProxySelector
    )
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    assertThat(routeSelector.hasNext()).isTrue
    dns[uriHost] = dns.allocate(1)
    val selection = routeSelector.next()
    assertRoute(selection.next(), address, Proxy.NO_PROXY, dns.lookup(uriHost, 0), uriPort)
    dns.assertRequests(uriHost)
    assertThat(selection.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isFalse
  }

  @Test fun proxySelectorReturnsNoProxies() {
    val address = httpAddress()
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    assertThat(routeSelector.hasNext()).isTrue
    dns[uriHost] = dns.allocate(2)
    val selection = routeSelector.next()
    assertRoute(selection.next(), address, Proxy.NO_PROXY, dns.lookup(uriHost, 0), uriPort)
    assertRoute(selection.next(), address, Proxy.NO_PROXY, dns.lookup(uriHost, 1), uriPort)
    assertThat(selection.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isFalse
    dns.assertRequests(uriHost)
    proxySelector.assertRequests(address.url.toUri())
  }

  @Test fun proxySelectorReturnsMultipleProxies() {
    val address = httpAddress()
    proxySelector.proxies.add(proxyA)
    proxySelector.proxies.add(proxyB)
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    proxySelector.assertRequests(address.url.toUri())

    // First try the IP addresses of the first proxy, in sequence.
    assertThat(routeSelector.hasNext()).isTrue
    dns[proxyAHost] = dns.allocate(2)
    val selection1 = routeSelector.next()
    assertRoute(selection1.next(), address, proxyA, dns.lookup(proxyAHost, 0), proxyAPort)
    assertRoute(selection1.next(), address, proxyA, dns.lookup(proxyAHost, 1), proxyAPort)
    dns.assertRequests(proxyAHost)
    assertThat(selection1.hasNext()).isFalse

    // Next try the IP address of the second proxy.
    assertThat(routeSelector.hasNext()).isTrue
    dns[proxyBHost] = dns.allocate(1)
    val selection2 = routeSelector.next()
    assertRoute(selection2.next(), address, proxyB, dns.lookup(proxyBHost, 0), proxyBPort)
    dns.assertRequests(proxyBHost)
    assertThat(selection2.hasNext()).isFalse

    // No more proxies to try.
    assertThat(routeSelector.hasNext()).isFalse
  }

  @Test fun proxySelectorDirectConnectionsAreSkipped() {
    val address = httpAddress()
    proxySelector.proxies.add(Proxy.NO_PROXY)
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    proxySelector.assertRequests(address.url.toUri())

    // Only the origin server will be attempted.
    assertThat(routeSelector.hasNext()).isTrue
    dns[uriHost] = dns.allocate(1)
    val selection = routeSelector.next()
    assertRoute(selection.next(), address, Proxy.NO_PROXY, dns.lookup(uriHost, 0), uriPort)
    dns.assertRequests(uriHost)
    assertThat(selection.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isFalse
  }

  @Test fun proxyDnsFailureContinuesToNextProxy() {
    val address = httpAddress()
    proxySelector.proxies.add(proxyA)
    proxySelector.proxies.add(proxyB)
    proxySelector.proxies.add(proxyA)
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    proxySelector.assertRequests(address.url.toUri())
    assertThat(routeSelector.hasNext()).isTrue
    dns[proxyAHost] = dns.allocate(1)
    val selection1 = routeSelector.next()
    assertRoute(selection1.next(), address, proxyA, dns.lookup(proxyAHost, 0), proxyAPort)
    dns.assertRequests(proxyAHost)
    assertThat(selection1.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isTrue
    dns.clear(proxyBHost)
    try {
      routeSelector.next()
      fail<Any>()
    } catch (expected: UnknownHostException) {
    }
    dns.assertRequests(proxyBHost)
    assertThat(routeSelector.hasNext()).isTrue
    dns[proxyAHost] = dns.allocate(1)
    val selection2 = routeSelector.next()
    assertRoute(selection2.next(), address, proxyA, dns.lookup(proxyAHost, 0), proxyAPort)
    dns.assertRequests(proxyAHost)
    assertThat(selection2.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isFalse
  }

  @Test fun multipleProxiesMultipleInetAddressesMultipleConfigurations() {
    val address = httpsAddress()
    proxySelector.proxies.add(proxyA)
    proxySelector.proxies.add(proxyB)
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)

    // Proxy A
    dns[proxyAHost] = dns.allocate(2)
    val selection1 = routeSelector.next()
    assertRoute(selection1.next(), address, proxyA, dns.lookup(proxyAHost, 0), proxyAPort)
    dns.assertRequests(proxyAHost)
    assertRoute(selection1.next(), address, proxyA, dns.lookup(proxyAHost, 1), proxyAPort)
    assertThat(selection1.hasNext()).isFalse

    // Proxy B
    dns[proxyBHost] = dns.allocate(2)
    val selection2 = routeSelector.next()
    assertRoute(selection2.next(), address, proxyB, dns.lookup(proxyBHost, 0), proxyBPort)
    dns.assertRequests(proxyBHost)
    assertRoute(selection2.next(), address, proxyB, dns.lookup(proxyBHost, 1), proxyBPort)
    assertThat(selection2.hasNext()).isFalse

    // No more proxies to attempt.
    assertThat(routeSelector.hasNext()).isFalse
  }

  @Test fun failedRouteWithSingleProxy() {
    val address = httpsAddress()
    var routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    val numberOfAddresses = 2
    dns[uriHost] = dns.allocate(numberOfAddresses)

    // Extract the regular sequence of routes from selector.
    val selection1 = routeSelector.next()
    val regularRoutes = selection1.routes

    // Check that we do indeed have more than one route.
    assertThat(regularRoutes.size).isEqualTo(numberOfAddresses)
    // Add first regular route as failed.
    routeDatabase.failed(regularRoutes[0])
    // Reset selector
    routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)

    // The first selection prioritizes the non-failed routes.
    val selection2 = routeSelector.next()
    assertThat(selection2.next()).isEqualTo(regularRoutes[1])
    assertThat(selection2.hasNext()).isFalse

    // The second selection will contain all failed routes.
    val selection3 = routeSelector.next()
    assertThat(selection3.next()).isEqualTo(regularRoutes[0])
    assertThat(selection3.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isFalse
  }

  @Test fun failedRouteWithMultipleProxies() {
    val address = httpsAddress()
    proxySelector.proxies.add(proxyA)
    proxySelector.proxies.add(proxyB)
    var routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    dns[proxyAHost] = dns.allocate(1)
    dns[proxyBHost] = dns.allocate(1)

    // Mark the ProxyA route as failed.
    val selection = routeSelector.next()
    dns.assertRequests(proxyAHost)
    val route = selection.next()
    assertRoute(route, address, proxyA, dns.lookup(proxyAHost, 0), proxyAPort)
    routeDatabase.failed(route)
    routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)

    // Confirm we enumerate both proxies, giving preference to the route from ProxyB.
    val selection2 = routeSelector.next()
    dns.assertRequests(proxyAHost, proxyBHost)
    assertRoute(selection2.next(), address, proxyB, dns.lookup(proxyBHost, 0), proxyBPort)
    assertThat(selection2.hasNext()).isFalse

    // Confirm the last selection contains the postponed route from ProxyA.
    val selection3 = routeSelector.next()
    dns.assertRequests()
    assertRoute(selection3.next(), address, proxyA, dns.lookup(proxyAHost, 0), proxyAPort)
    assertThat(selection3.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isFalse
  }

  @Test fun queryForAllSelectedRoutes() {
    val address = httpAddress()
    val routeSelector = RouteSelector(address, routeDatabase, call, EventListener.NONE)
    dns[uriHost] = dns.allocate(2)
    val selection = routeSelector.next()
    dns.assertRequests(uriHost)
    val routes = selection.routes
    assertRoute(routes[0], address, Proxy.NO_PROXY, dns.lookup(uriHost, 0), uriPort)
    assertRoute(routes[1], address, Proxy.NO_PROXY, dns.lookup(uriHost, 1), uriPort)
    assertThat(selection.next()).isSameAs(routes[0])
    assertThat(selection.next()).isSameAs(routes[1])
    assertThat(selection.hasNext()).isFalse
    assertThat(routeSelector.hasNext()).isFalse
  }

  @Test fun getHostString() {
    // Name proxy specification.
    var socketAddress = InetSocketAddress.createUnresolved("host", 1234)
    assertThat(socketAddress.socketHost).isEqualTo("host")
    socketAddress = InetSocketAddress.createUnresolved("127.0.0.1", 1234)
    assertThat(socketAddress.socketHost).isEqualTo("127.0.0.1")

    // InetAddress proxy specification.
    socketAddress = InetSocketAddress(InetAddress.getByName("localhost"), 1234)
    assertThat(socketAddress.socketHost).isEqualTo("127.0.0.1")
    socketAddress = InetSocketAddress(InetAddress.getByAddress(byteArrayOf(127, 0, 0, 1)), 1234)
    assertThat(socketAddress.socketHost).isEqualTo("127.0.0.1")
    socketAddress = InetSocketAddress(
      InetAddress.getByAddress("foobar", byteArrayOf(127, 0, 0, 1)),
      1234
    )
    assertThat(socketAddress.socketHost).isEqualTo("127.0.0.1")
  }

  @Test fun routeToString() {
    val route = Route(
      httpAddress(),
      Proxy.NO_PROXY,
      InetSocketAddress.createUnresolved("host", 1234)
    )
    val expected = when {
      isAndroid || majorVersion < 14 -> "Route{host:1234}"
      else -> "Route{host/<unresolved>:1234}"
    }
    assertThat(route.toString()).isEqualTo(expected)
  }

  private fun assertRoute(
    route: Route,
    address: Address,
    proxy: Proxy,
    socketAddress: InetAddress,
    socketPort: Int
  ) {
    assertThat(route.address).isEqualTo(address)
    assertThat(route.proxy).isEqualTo(proxy)
    assertThat(route.socketAddress.address).isEqualTo(socketAddress)
    assertThat(route.socketAddress.port).isEqualTo(socketPort)
  }

  /** Returns an address that's without an SSL socket factory or hostname verifier.  */
  private fun httpAddress(): Address {
    return Address(
      uriHost = uriHost,
      uriPort = uriPort,
      dns = dns,
      socketFactory = socketFactory,
      sslSocketFactory = null,
      hostnameVerifier = null,
      certificatePinner = null,
      proxyAuthenticator = authenticator,
      proxy = null,
      protocols = protocols,
      connectionSpecs = connectionSpecs,
      proxySelector = proxySelector
    )
  }

  private fun httpsAddress(): Address {
    return Address(
      uriHost = uriHost,
      uriPort = uriPort,
      dns = dns,
      socketFactory = socketFactory,
      sslSocketFactory = sslSocketFactory,
      hostnameVerifier = hostnameVerifier,
      certificatePinner = null,
      proxyAuthenticator = authenticator,
      proxy = null,
      protocols = protocols,
      connectionSpecs = connectionSpecs,
      proxySelector = proxySelector
    )
  }

  companion object {
    private const val proxyAPort = 1001
    private const val proxyAHost = "proxya"
    private val proxyA = Proxy(
      Proxy.Type.HTTP,
      InetSocketAddress.createUnresolved(proxyAHost, proxyAPort)
    )

    private const val proxyBPort = 1002
    private const val proxyBHost = "proxyb"
    private val proxyB = Proxy(
      Proxy.Type.HTTP,
      InetSocketAddress.createUnresolved(proxyBHost, proxyBPort)
    )
  }
}
